### ISSUE TITLE
RNN fails to forward in GPU mode

### ISSUE THREAD

**Issue by NightFury13** [Element-Research/rnn#277]
I am trying to prototype the pipeline for running a RNN on GPU. Following is my model.
```
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> output]
  (1): nn.Sequencer @ nn.FastLSTM(32 -> 50)
  (2): nn.Sequencer @ nn.Recursor @ nn.Linear(50 -> 109)
  (3): nn.Sequencer @ nn.Recursor @ nn.LogSoftMax
}
```
I feed it with a random input created as follows.
```
input = {}
for i=1,50 do
    input[i] = torch.Tensor(32):random(10); 
    input[i]=input[i]:cuda()
end
```
And I also covert the rnn to GPU mode using, `cutorch.setDevice(1); net = net:cuda()`.

Now when I forward the model with the inputs, `net:forward(input)`, I get the following error trace.
```
THCudaCheck FAIL file=/tmp/luarocks_cutorch-scm-1-7719/cutorch/lib/THC/generic/THCTensorMathPointwise.cu line=40 error=8 : invalid device function
.../mohit.jain/torch/install/share/lua/5.1/nn/Container.lua:67: 
In 1 module of nn.Sequential:
In 1 module of nn.Sequential:
In 1 module of nn.ConcatTable:
In 3 module of nn.Sequential:
.../mohit.jain/torch/install/share/lua/5.1/nn/CAddTable.lua:16: cuda runtime error (8) : invalid device function at /tmp/luarocks_cutorch-scm-1-7719/cutorch/lib/THC/generic/THCTensorMathPointwise.cu:40
stack traceback:
	[C]: in function 'add'
	.../mohit.jain/torch/install/share/lua/5.1/nn/CAddTable.lua:16: in function <.../mohit.jain/torch/install/share/lua/5.1/nn/CAddTable.lua:9>
	[C]: in function 'xpcall'
	.../mohit.jain/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...mohit.jain/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function <...mohit.jain/torch/install/share/lua/5.1/nn/Sequential.lua:41>
	[C]: in function 'xpcall'
	.../mohit.jain/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...ohit.jain/torch/install/share/lua/5.1/nn/ConcatTable.lua:11: in function <...ohit.jain/torch/install/share/lua/5.1/nn/ConcatTable.lua:9>
	[C]: in function 'xpcall'
	.../mohit.jain/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...mohit.jain/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/users/mohit.jain/torch/install/share/lua/5.1/rnn/LSTM.lua:162: in function 'updateOutput'
	...mohit.jain/torch/install/share/lua/5.1/rnn/Sequencer.lua:58: in function <...mohit.jain/torch/install/share/lua/5.1/rnn/Sequencer.lua:39>
	[C]: in function 'xpcall'
	.../mohit.jain/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...mohit.jain/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	[string "_RESULT={net:forward(input)}"]:1: in main chunk
	[C]: in function 'xpcall'
	...rs/mohit.jain/torch/install/share/lua/5.1/trepl/init.lua:651: in function 'repl'
	...jain/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
	[C]: at 0x00406670

WARNING: If you see a stack trace below, it doesn't point to the place where this error occured. Please use only the one above.
stack traceback:
	[C]: in function 'error'
	.../mohit.jain/torch/install/share/lua/5.1/nn/Container.lua:67: in function 'rethrowErrors'
	...mohit.jain/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	[string "_RESULT={net:forward(input)}"]:1: in main chunk
	[C]: in function 'xpcall'
	...rs/mohit.jain/torch/install/share/lua/5.1/trepl/init.lua:651: in function 'repl'
	...jain/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
	[C]: at 0x00406670
```

Am I converting the model/data incorrectly (the corresponding CPU version works just fine), or is there a problem with my GPU/library environments?

FYI, output of nvidia-smi
```
+------------------------------------------------------+                       
| NVIDIA-SMI 352.39     Driver Version: 352.39         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K40c          Off  | 0000:01:00.0     Off |                    0 |
| 27%   55C    P0    62W / 235W |    541MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      2492    C   python2                                        339MiB |
|    0      8673    C   python2                                         87MiB |
|    0     18114    C   python2                                         87MiB |
+-----------------------------------------------------------------------------+
```

**Comment by NightFury13** [comment]
Not sure if this is the most elegant solution, but here's what solved it for me.

I was using a different GPU (from the cluster I access) from the one which I compiled my torch cunn and cutorch libraries on. By simply re-installing these libraries with luarocks, the errors are gone. Hope this helps someone else stuck in a similar problem.

**Comment by nicholas-leonard** [comment]
@NightFury13 Yeah I think that the issue would also have been resolved by calling cutorch.setDevice before casting anything (input or model) to cuda.

**Comment by NightFury13** [comment]
@nicholas-leonard : Actually the error popped up irrespective of where I called the `cutorch.setDevice` routine. I generally place this method at the very beginning of my scripts so that shouldn't have been a problem. I'm guessing its got something to do with the binaries that are built using `luarocks install cunn/cutorch`. Since I shifted the underlying architecture that I was running my code on (Multiple GPUs in a cluster-of-machines scenario, I changed the node on which my code ran) these binaries probably needed to be built again with the configuration of the new machine.
